### PR TITLE
Enhance EMFExporter for Metrics Batching in AWS EMF Logs 

### PR DESCRIPTION
**Pull request by mxiamxia**
**Description:** 
We sent a large PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1891 to support batching the metrics on the same dimensions for AWS EMF Log request to save the customers' billing cost and request throughput. At the same time, there was a fairly large code refactor on EMFExporter. For better code review purpose, I plan to split https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1891  to 2 PRs. (This is PR#1)

In this PR, We refactored EMFExporter without introducing any new feature. For each OTel metrics data point, we defined `DataPoint` file, it wraps `pdata.DataPointSlice` to the custom structures for each type of metric data point. we also moved the metric data handling functions - data conversion and rate calculation to `datapoint`. 
It also fixed the metric `timestamp` bug.

### CODE REVIEW

**Comment by mxiamxia** [comment]
@anuraaga Thanks!

File: exporter/awsemfexporter/datapoint.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	return DataPoint{
+		Value:       metricVal,
+		Labels:      labels,
+		TimestampMs: timestampMs,
+	}
+}
+
+// createLabels converts OTel StringMap labels to a map and optionally adds in the
```
Comment (anuraaga) [inline exporter/awsemfexporter/datapoint.go:184]: What is this "optional OTel instrumentation library name"?

File: exporter/awsemfexporter/datapoint.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	return metricRate
+}
+
+// getDataPoints retrieves data points from OT Metric.
+func getDataPoints(pmd *pdata.Metric, metadata CWMetricMetadata, logger *zap.Logger) (dps DataPoints) {
+	if pmd == nil {
+		return
+	}
+
+	rateKeyParams := map[string]string{
```
Comment (anuraaga) [inline exporter/awsemfexporter/datapoint.go:224]: Looks like this should be a struct, not map

File: exporter/awsemfexporter/util.go [hunk 3]
```diff
@@ -55,3 +60,72 @@ func replace(s, pattern string, value pdata.AttributeValue, logger *zap.Logger)
...
+		logGroup = replacePatterns(config.LogGroupName, rm.Resource().Attributes(), config.logger)
+	}
+	if len(config.LogStreamName) > 0 {
+		logStream = replacePatterns(config.LogStreamName, rm.Resource().Attributes(), config.logger)
+	}
+
+	return
+}
+
+// createMetricKey generates a hashed key from metric labels and additional parameters
```
Comment (anuraaga) [inline exporter/awsemfexporter/util.go:101]: So it doesn't necessarily have to be in this PR but I do think we need to update this to use `label.Distinct` as we did in the statsdreceiver as its specially optimized for this use case

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1891#discussion_r551750484

Then we'll be able to use just a normal struct with the current `rateKeyParams` and a `label.Distinct`

File: exporter/awsemfexporter/metric_translator.go [hunk 4]
```diff
@@ -77,75 +58,23 @@ type CWMetricStats struct {
...
-			namespace = serviceNamespace.StringVal()
-		}
-	}
-
-	if len(namespace) == 0 {
-		namespace = defaultNameSpace
-	}
+	cWNamespace := getNamespace(rm, config.Namespace)
+	logGroup, logStream := getLogInfo(rm, cWNamespace, config)
+	timestampMs := time.Now().UnixNano() / int64(time.Millisecond)
```
Comment (anuraaga) [inline exporter/awsemfexporter/metric_translator.go:77]: Does `time.Now() / time.Millisecond` work?

File: exporter/awsemfexporter/datapoint_test.go [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
```
Comment (anuraaga) [inline exporter/awsemfexporter/datapoint_test.go:22]: Are we still using opencensus somewhere?

**Review by anuraaga — COMMENTED** [review]
Thanks!

File: exporter/awsemfexporter/datapoint.go (hunk 1 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/datapoint.go:184]: I should remove this line of comment. It was introduced in the closed PR(large one) which needs to use `instrumentation library` as key for grouping the metrics. `OTel instrumentation library dimension` is still handled in `metric_translator::buildCWMetric` function. I will add it back in next PR.

File: exporter/awsemfexporter/util.go (hunk 3 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/util.go:101]: Thanks, I'll take a look the `label.Distinct`. We need to use `createMetricKey` for counter metric rate calculation in this PR as part of refactor.

File: exporter/awsemfexporter/metric_translator.go (hunk 4 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/metric_translator.go:77]: it looks not, `time.Now()` returns `Time` struct that can't be divided, and it seems `timestampMs := time.Now().UnixNano() / int64(time.Millisecond)` is pretty common practice for getting current timestamp in MS. :)

File: exporter/awsemfexporter/datapoint_test.go (hunk 5 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/datapoint_test.go:22]: Yeah, we still reply on `opencensus` and `internaldata.OCToMetrics(oc)` to generate OTel Metrics test data in some of  metrics exporter test cases. We'll create a specific PR for the test data migration.

File: exporter/awsemfexporter/util.go (hunk 3 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/util.go:101]: Hi @anuraaga, I tried to follow the idea that use a structure to define `rateKeyParams` fields along with `label.Distinct` that converted from `metric.LabelMap`, and then generate the metric rate calculation key string form both. The code changes had looked weird to me.  In my case, I have to convert the metric lablels `pdata.Metric.LabelMap` to `[]lable.KeyValue` so I can use `label.Distinct` sortting my labels. After that, I have to parse `label.Distinct`and `rateKeyParams` key/value pairs to calculate the rate key string like "paramkey:paramValue, labelkey:value,labelKey2:value2" which I think it is much complicated than the current `createMetricKey` implementation.  I feel this is not fitting my case unless I am using `label.Distinct` in the wrong way.  

I can convert the current `rateKeyParams` map to structure which makes sense to me, but not sure how I get the help with `label.Distinct` for my case.

Comment (anuraaga) [reply exporter/awsemfexporter/util.go:101]: > After that, I have to parse label.Distinctand rateKeyParams key/value pairs to calculate the rate key string like "paramkey:paramValue, labelkey:value,labelKey2:value2"

I don't think you should need to compute a string anymore - once you have a struct with all the fields you need, including a `label.Distinct`, that struct could be the map key, not string, right?

File: exporter/awsemfexporter/datapoint_test.go (hunk 5 above)
Comment (anuraaga) [reply exporter/awsemfexporter/datapoint_test.go:22]: I'm a bit worried about going 1.0 with this, what if the metrics change drastically? I guess not but we need to confirm this is safe, or migrate soon. We probably were supposed to have been migrated a while ago really.

File: exporter/awsemfexporter/util.go (hunk 3 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/util.go:101]: > that struct could be the map key, not string, right?

Comment (anuraaga) [reply exporter/awsemfexporter/util.go:101]: Yeah - we already have it as a dep in the statsd receiver so I guess it's ok?

File: exporter/awsemfexporter/datapoint_test.go (hunk 5 above)
Comment (mxiamxia) [reply exporter/awsemfexporter/datapoint_test.go:22]: I see. It'll be a problem If no one maintains `internaldata.OCToMetrics(oc)` for the new Metrics updates. I see there are a few metrics components are creating test data in this way for unit tests.  I can create an issue for it to call out the problem.

**Review by anuraaga — APPROVED** [review]

**Comment by mxiamxia** [comment]
The `build-publish` failures are not related to the changes in this PR.

**Comment by mxiamxia** [comment]
Hi @tigrannajaryan @bogdandrutu , Please help to review and merge the changes. Thanks.

**Comment by tigrannajaryan** [comment]
Approved by codeowner, merging.

**Pull request merged**